feature/hack-3 (#587)
hirosystems/docs: docs/hacks/build-a-custom-api-using-chainhook.mdx

### Creating a Chainhook

Let's create our first Chainhook. We're going to start by creating a predicate that matches the `propose` function in your `proposal-submission` contract. This function is called when a new grant proposal is submitted.

Here's an example of how you might define this `predicate.json` file for your contract:

```json
{
  "chain": "stacks",
  "uuid": "1",
  "name": "New Grant Proposal",
  "version": 1,
  "networks": {
    "testnet": {
      "if_this": {
        "scope": "contract_call",
        "contract_identifier": "ST2BSV94A650WGZ2YZ5Y8HM93W01NGT4GY0MGJECG.proposal-submission",
        "method": "propose"
      },
      "then_that": {
        "http_post": {
          "url": "<your-https-server-url>",
          "authorization_header": "Bearer 12345"
        }
      },
      "start_block": 138339
    }
  }
}
```
> _**Note**_
>

> > Make sure to swap out the details in your `predicate.json` file to match your contracts ABI

Reviewer: ```suggestion
> Make sure to swap out the details in your `predicate.json` file to match your contracts ABI.
```

> The `if_this` section specifies the conditions for the events that this Chainhook will respond to. In this case, it's looking for calls to the `propose` method in the `proposal-submission` contract on the Stacks testnet.

Reviewer: ```suggestion
The `if_this` section specifies the conditions for the events that this chainhook will respond to. In this case, it's looking for calls to the `propose` method in the `proposal-submission` contract on the Stacks testnet.
```

> The `then_that` section specifies what action the Chainhook should take when it detects an event that matches the `if_this` conditions. Here, it's set up to send a POST request to your specified URL with the event data.

Reviewer: ```suggestion
The `then_that` section specifies what action the chainhook should take when it detects an event that matches the `if_this` conditions. Here, it's set up to send a POST request to your specified URL with the event data.
```

> The `start_block` field is used to specify the starting block for the Chainhook to start listening from. This is useful for ignoring blocks that were mined before the Chainhook was set up.

Reviewer: ```suggestion
The `start_block` field is used to specify the starting block for the chainhook to start listening from. This is useful for ignoring blocks that were mined before the chainhook was set up.
```

Once you have created a similar file for your specific contract. select the _Chainhooks_ tab inside your project and upload your predicate file using the *Upload Chainhook* option. For more information, you can follow the [Create Chainhooks](../platform/guides/create-chainhooks) section. 

## Testing Your Chainhook

> Once you've successfully uploaded your Chainhook predicate, it's time to test it out. You can do this by performing an action that aligns with your `if_this` - `then_that` logic.

Reviewer: ```suggestion
Once you've successfully uploaded your chainhook predicate, it's time to test it out. You can do this by performing an action that aligns with your `if_this` - `then_that` logic.
```

> For instance, if your Chainhook is set up to respond to a specific contract call, you can trigger that call and then check your `/api/post` endpoint. If everything is set up correctly, you should see any information you've requested to be logged.

Reviewer: ```suggestion
For instance, if your chainhook is set up to respond to a specific contract call, you can trigger that call and then check your `/api/post` endpoint. If everything is set up correctly, you should see any information you've requested to be logged at that endpoint.
```

Congratulations on reaching this point! You now have all the building blocks needed to create a custom API using Chainhook and integrate it with your smart contracts. This is a significant step towards enhancing the functionality and user experience of your decentralized applications.

The next section presents several challenges that will help you put the finishing touches on your API. With the foundation you've built, you're well-equipped to take on these challenges and continue your journey in building robust, decentralized applications.

> > _**Note**_

Reviewer: this is an odd place for this note. can we move this two paragraphs up, so it doesn't follow the congrats/next steps paragraph? Imo this should come right after "If everything is set up correctly, you should see any information you've requested to be logged."

>
> This is a crucial step in verifying that your Chainhook is working as expected. If you don't see the expected output, you may need to revisit your predicate and ensure it's correctly configured.

## Challenges

The following challenges are additional features you can implement to further differentiate your project. Feel free to add any other features you want.

<span className="inline-flex items-center rounded-md bg-green-50 dark:bg-green-400/10 px-2 py-1 mb-2 text-xs font-medium text-green-700 dark:text-green-400 ring-1 ring-inset ring-green-600/10 dark:ring-green-400/20">
  Starter
</span>

> **Create more than one Chainhook**: Don't stop at just one! Follow the process from above to create more than one Chainhook for your dApp and parse the data accordingly.

Reviewer: ```suggestion
**Create More Than 1 Chainhook**: Don't stop at just 1! Follow the process above to create more than 1 chainhook for your dApp and parse the data accordingly.
```

<span className="inline-flex items-center rounded-md bg-yellow-50 dark:bg-yellow-400/10 px-2 py-1 mb-2 text-xs font-medium text-yellow-700 dark:text-yellow-400 ring-1 ring-inset ring-yellow-600/10 dark:ring-yellow-400/20">
  Intermediate
</span>

> **Create Consumer-Facing API Endpoints**: Up until now, we've shown you how to capture and process the initial `payload` data sent from Chainhook. However, to provide a robust user experience, you'll want to do more than just output raw data. The challenge here is to create additional API endpoints that present the filtered and extracted data from the initial `/api/events` payload in a consumer-friendly format. This could involve adding a database layer for organizing the data, adding additional context, or transforming the data to match the needs of your application's frontend. For a quick db setup, you can try using something like [Supabase](https://supabase.com/docs/guides/getting-started/quickstarts/reactjs).

Reviewer: ```suggestion
**Create Consumer-Facing API Endpoints**: Up until now, we've shown you how to capture and process the initial `payload` data sent from a chainhook. However, to provide a robust user experience, you'll want to do more than just output raw data. The challenge here is to create additional API endpoints that present the filtered and extracted data from the initial `/api/events` payload in a consumer-friendly format. This could involve adding a database layer for organizing the data, adding additional context, or transforming the data to match the needs of your application's frontend. For a quick db setup, you can try using something like [Supabase](https://supabase.com/docs/guides/getting-started/quickstarts/reactjs).
```

app.post("/api/events", async (req, res) => {
  // ...

    item.transactions.forEach((transaction: any) => {
      if (transaction.operations) {
        transaction.operations.forEach((operation: any) => {
          const data = operation // Parse out the data you want
          // Insert and save it to db
          const { error } = await supabase
            .from('proposals')
            .insert(data)
          });
      }
    });

  // ...
});
```

<span className="inline-flex items-center rounded-md bg-orange-50 dark:bg-orange-400/10 px-2 py-1 mb-2 text-xs font-medium text-orange-700 dark:text-orange-400 ring-1 ring-inset ring-orange-600/10 dark:ring-orange-400/20">
  Advanced
</span>

**UI Integration**: The final challenge is to integrate your consumer-facing API into your frontend. This involves taking the data you've processed and organized in your API and displaying it in a meaningful and user-friendly way in your application. This could involve creating dynamic components that update in response to new data, adding interactive elements that allow users to explore the data, or incorporating visualizations to help users understand the data.

> 

Reviewer: ```suggestion
## Prizes and Submissions

We are giving 300K sats each to the 10 best submissions for this hack. If you've participated in the last 2 hacks and complete this one too, we will also be giving a grand prize of 3M sats to the dev who submits the best and most consistent projects across all 3 hacks.

In order to be eligible for both of these prizes, you must complete the hack and submit before the deadline. The deadline for this hack is next Wednesday, December 20th at midnight ET.

To submit your project, [please use this form](https://survey.hiro.so/hirohacks). Good luck!

View the official Hiro Hacks rules [here](https://www.hiro.so/hiro-hacks-rules). 
```

> <div>

Reviewer: this formatting is wonky so I opted to rewrite the copy in markdown above. everything below this comment can be deleted
